README.md
# GCM

The library was inspired by [apns4ex](https://github.com/chvanikoff/apns4ex) and
much of the source code has been used as a bootstrap for this lib.

The HTTP wrapper has been heavily inspired by [carnivalmobile/gcm](https://github.com/carnivalmobile/gcm).

## Warning

This library is a work in progress and it's API is subject to change till `v0.1`, please consider use of `== ver` operator rather than `~> ver` when requiring `gcm4ex` as a dependency or your application may be broken with next release of the library.

## Installation

  1. Add gcm to your list of dependencies in mix.exs:

        def deps do
          [{:gcm, "== 0.0.2"}]
        end

  2. Ensure gcm is started before your application:

        def application do
          [applications: [:gcm]]
        end

## Usage

Config the GCM app and define pools

```elixir
config :gcm,
  # Here are pools configs. Any value from "global" config can be overwritten in any single pool config
  success_callback_module: GCM.Callbacks.SuccessHandler,
  error_callback_module: GCM.Callbacks.ErrorHandler,
  batch_size: 1000,
  pools: [
    # app1_dev_pool is the pool_name
    app1_dev_pool: [
      pool_size: 10,
      pool_max_overflow: 5,
      # and this is overwritten config key
      key: "my gcm api key"
    ],
    app1_prod_pool: [
      key: "my gcm api key"
      pool_size: 100,
      pool_max_overflow: 50
    ],
  ]
```

### Config keys

| Name                    | Default value                | Description                                                      |
|:------------------------|:-----------------------------|:-----------------------------------------------------------------|
| success_callback_module | GCM.Callbacks.SuccessHandler | This module receive request and response data on success         |
| error_callback_module*  | GCM.Callbacks.ErrorHandler   | This module receive request and response data on error           |
| batch_size              | 1000                         | Number of registration ids to send with each multicast request** |
| pools                   | []                           | List of pools to start                                           |

\* Google do not allow more than 1000 registration ids to be sent in the same request!

### Pool keys

| Pool key          | Description                                        |
|:------------------|:---------------------------------------------------|
| key               | GCM API key                                        |
| pool_size         | Maximum pool size                                  |
| pool_max_overflow | Maximum number of workers created if pool is empty |

All pools defined in config will be started automatically

From here and now you can start pushing your PNs via GCM.push/2 and GCM.push/3

### Handle responses

The library comes with default callback modules that do logging. You might be
fine with the default `error_callback_module` but you'd want to write your own
`success_callback_module`. It need to handle the following response keys by updating
your database.

| Name                        | Action | Example                                      |
|:----------------------------|:-------|:---------------------------------------------|
| canonical_ids               | update | [%{old: "reg1", new: "newreg1"}]             |
| invalid_registration_ids    | delete | invalid_registration_ids: ["reg2"]           |
| not_registered_ids          | delete | not_registered_ids: ["reg3"]                 |
| deletable_registration_ids* | delete | deletable_registration_ids: ["reg2", "reg3"] |

\* `deletable_registration_ids` is a concatenation of `invalid_registration_ids` and `not_registered_ids`.
There should not be duplicates but it's not guaranteed by the lib. If you don't need
to distinguish between invalid and not registered ids you can ignore these keys and
only use `deletable_registration_ids`.


```elixir
[{:ok, %{
  status_code: 200,
  success: 2,
  failure: 0,
  body: "{}",
  canonical_ids: [],
  invalid_registration_ids: [],
  not_registered_ids: [],
  headers: [{"Content-Type", "application/json; charset=UTF-8"}, …]
}}, …]
```

See https://developers.google.com/cloud-messaging/http for more info

## Basic Usage

### Example using the notification payload:

```elixir
message =
  GCM.Message.new
  |> Map.put(:notification, %GCM.Message.Notification{title: "Hello world!"})
  |> Map.put(:data, %{"post_id" => "23"})

GCM.push(:app1_dev_pool, message)
```

### Example using the custom data structs:

```elixir
message = Map.put(GCM.Message.new, :data, %{"my-custom-key" => "Hello world!"})
GCM.push(:app1_dev_pool, message)
```

### Send un-supervised, one-off push messages

The supervised, pooled `GCM.push\2` is probably what you want to use in your app
but if you just want to play around with push messages from the console it can be
convenient to use the bare `GCM.Sender.push` function:

A successful push looks like this:

```
iex> GCM.Sender.push("api_key", ["registration_id1", "registration_id2"], %{notification: %{title: "Hello!"}})
```

If the push failed the return is `{:error, reason}` where reason will include more information on what failed.

More info here: https://developers.google.com/cloud-messaging/http


CHANGELOG.md
# Changelog

## 0.0.3 (2016-03-11)

Add deletable_registration_ids response key to make it more convenient to
delete invalid or not registered registration ids.

## 0.0.2 (2016-03-08)

Add support for splitting multipart requests into batches as required by Google.

## 0.0.1 (2016-03-04)

* Initial release, heavily inspired by [apns4ex](https://github.com/chvanikoff/apns4ex) and [carnivalmobile/gcm](https://github.com/carnivalmobile/gcm).


lib/gcm/sender.ex
# Stolen from https://github.com/carnivalmobile/gcm

defmodule GCM.Sender do
  alias HTTPoison.Response
  require Logger

  @base_url "https://gcm-http.googleapis.com/gcm"
  @url @base_url <> "/send"
  @empty_results %{not_registered_ids: [], canonical_ids: [], invalid_registration_ids: [], deletable_registration_ids: []}
  @batch_size Application.get_env(:gcm, :batch_size) || 1000
  @success_callback_module Application.get_env(:gcm, :success_callback_module) || GCM.Callbacks.SuccessHandler
  @error_callback_module Application.get_env(:gcm, :error_callback_module) || GCM.Callbacks.ErrorHandler

  def push(api_key, registration_ids, payload \\ %{}, http_module \\ HTTPoison) do
    registration_ids = List.wrap(registration_ids)
    headers = headers(api_key)

    Enum.map(Enum.chunk(registration_ids, @batch_size, @batch_size, []), fn (registration_ids_chunk) ->
      send_request(registration_ids_chunk, headers, payload, http_module)
    end)
  end

  defp send_request(registration_ids, headers, payload, http_module) do
    body = case registration_ids do
      [id] -> %{to: id}
      ids -> %{registration_ids: ids}
    end |> Dict.merge(payload)

    case http_module.post(@url, Poison.encode!(body), headers) do
      {:ok, response} ->
        case build_response(registration_ids, response) do
          {:ok, response} ->
            handle_success(request: %{url: @url, body: body, headers: headers}, response: response)
          {:error, reason} ->
            handle_error(request: %{url: @url, body: body, headers: headers}, reason: reason)
        end
      {:error, reason} ->
        handle_error(request: %{url: @url, body: body, headers: headers}, reason: reason)
    end
  end

  defp handle_success(request: request, response: response) do
    @success_callback_module.handle(request: request, response: response)
    {:ok, response}
  end

  defp handle_error(request: request, reason: reason) do
    @error_callback_module.handle(request: request, reason: reason)
    {:error, reason}
  end

  defp build_response(_, %Response{status_code: 400}), do: {:error, :bad_request}
  defp build_response(_, %Response{status_code: 401}), do: {:error, :unauthorized}
  defp build_response(_, %Response{status_code: 503}), do: {:error, :service_unavailable}
  defp build_response(_, %Response{status_code: code}) when code in 500..599, do: {:error, :server_error}
  defp build_response(registration_ids, %Response{headers: headers, status_code: 200, body: body}) do
    response = Poison.decode!(body)
    results = build_results(response, registration_ids)
    defaults = %{
      failure: response["failure"],
      success: response["success"],
      body: body,
      headers: headers,
      status_code: 200,
      deletable_registration_ids: results[:not_registered_ids] ++ results[:invalid_registration_ids]
    }

    {:ok, Map.merge(results, defaults)}
  end

  defp build_results(%{"failure" => 0, "canonical_ids" => 0}, _), do: @empty_results
  defp build_results(%{"results" => results}, reg_ids) do
    response = @empty_results

    Enum.zip(reg_ids, results)
    |> Enum.reduce response, fn({reg_id, result}, response) ->
      case result do
        %{"error" => "NotRegistered"} ->
          update_in(response[:not_registered_ids], &([reg_id | &1]))
        %{"error" => "InvalidRegistration"} ->
          update_in(response[:invalid_registration_ids], &([reg_id | &1]))
        %{"registration_id" => new_reg_id} ->
          update = %{old: reg_id, new: new_reg_id}
          update_in(response[:canonical_ids], &([update | &1]))
        _ -> response
      end
    end
  end
  defp build_results(_, _), do: @empty_results

  defp headers(api_key) do
    [{ "Authorization", "key=#{api_key}" },
     { "Content-Type", "application/json" },
     { "Accept", "application/json"}]
  end
end
